text/2018-10-25-batch-split.md

```diff
@@ -0,0 +1,212 @@
+    metapb.Region region = 2;
+    uint32 split_count = 3;
+}
+
+message SplitID {
+    uint64 new_region_id = 1;
+    repeated uint64 new_peer_ids = 2;
+}
+
+message AskBatchSplitResponse {
+    ResponseHeader header = 1;
+    repeated SplitID ids = 2;
+}
+
+message ReportBatchSplitRequest {
+    RequestHeader header = 1;
+    repeated metapb.Region regions = 2;
+}
+
+message ReportBatchSplitResponse {
+    ResponseHeader header = 1;
+}
+```
+
+Add `AskBatchSplit` to replace `AskSplit`. It is called when TiKV produces some 
+split keys for one Region and asks PD to allocate new `region_id` and `peer_id` 
+for that Region. `split_count` in `AskBatchSplitRequest` indicates the number 
+of Regions to be generated, and `AskBatchSplitResponse` returns all new 
+allocated IDs to TiKV.
+
+Add `ReportBatchSplit` to replace `ReportBatchSplit`. It is called when TiKV 
+finishes splitting a Region. `ReportBatchSplitRequest` takes all metas of a new 
+generated Region for PD to update PD's related information.
+
+For compatibility issue, the old interface is not deleted but set to 
+deprecated. 
+
+### TiKV
+
+```protobuf
+message SplitRequest {
+    // ...
+    // Will be ignored in batch split. Use `BatchSplitRequest::right_derive` instead.
+    bool right_derive = 4 [deprecated=true];
+}
+
+message BatchSplitRequest {
+    repeated SplitRequest requests = 1;
+    // If true, the last Region obtains the origin region_id, 
+    // and other Regions use new Ids.
+    bool right_derive = 2;
+}
+
+message BatchSplitResponse {
+    repeated metapb.Region regions = 1;
+}
+
+enum AdminCmdType {
+    // ...
+    Split = 2 [deprecated=true];
+    // ...
+    BatchSplit = 10;
+}
+
+message AdminRequest {
+    // ...
+    SplitRequest split = 3 [deprecated=true];
+    // ...
+    BatchSplitRequest splits = 10;
+}
+
+message AdminResponse {
+    // ...
+    SplitResponse split = 3 [deprecated=true];
+    // ...
+    BatchSplitResponse splits = 10;
+}
+```
+
+Add a new admin command type `BatchSplit` with related request and response. 
+`BatchSplitRequest` wraps multiple `SplitRequest` along with `right_derive` 
+which invalidates the `right_derive` in each `SplitRequest`.
+
+When in the rolling upgrade process, new TiKVs are mixed up with old TiKVs, so 
+old command type `Split` still needs to be preserved.
+
+## Implementation in TiKV
+
+### How to produce multiple split keys
+
+First introducing one config `batch_split_limit` to limit the number of produced 
+split keys in a batch. If it is unlimited, for a once split check, it scans all 
+over the Region's range, and in some extreme case, this would cause performance 
+issue.
+
+SizeChecker and KeysChecker can be rewritten to produce multiple keys, and the 
+general logic of SizeChecker and KeysChecker are similar. The only difference 
+between them is one splits Region based on size and the other splits Region 
+based on the number of keys. So here we mainly describe the logic of 
+SizeChecker:
```

Comment:
Actually, I think the following two paragraphs are too verbose and not necessary.
These explanations are more suitable as comments in the code than in the RFC.

Reply:
IMO, the algorithm should discussed in RFC stage.